### DISCUSSION TITLE
Add BodyOrQuery param type

### DISCUSSION THREAD

**Discussion by falkben** [fastapi/fastapi#9113] [Questions]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * Or, I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

```Python
from typing import Optional
import uvicorn
from fastapi import FastAPI, Depends
from pydantic import BaseModel

app = FastAPI()


class Item(BaseModel):
    p_int: Optional[int] = None
    p_float: Optional[float] = None
    p_str: Optional[str] = None


class BodyOrQuery:
    def __init__(self, item: Item = None, query: Item = Depends()):
        # merging here

        item_dict = {}
        if item is not None:
            item_dict = item.dict()

        self.items = {
            **{k: v for k, v in item_dict.items() if v is not None},
            **{k: v for k, v in query.dict().items() if v is not None},
        }


@app.post("/")
def body_or_query_using_depends(body_or_query: BodyOrQuery = Depends()):
    return body_or_query.items
```

### Description

There is a related issue (https://github.com/tiangolo/fastapi/issues/1543) but that was a question and this is a feature request.

Tornado supports [getting arguments](https://www.tornadoweb.org/en/stable/web.html#tornado.web.RequestHandler.get_arguments) from the body or query params.  We're trying to convert a tornado app to fastapi and will need to support this existing behavior.  My workaround above shows one way we could possibly support body or query arguments in the same method, but it has the drawback of making both sets of argument types optional and ends up being a lot of code to write for each path operator.  Also, if in the example above, `p_int` was required, that would not show up in swagger or pydantic, and we'd have to do our own validation.

### The solution you would like

I would like it to have a `BodyOrQuery` parameter type that could take arguments from either input:

```Python
from fastapi import FastAPI

app = FastAPI()


@app.post("/")
def body_or_query(p_int: int = BodyOrQuery(), p_float: float = BodyOrQuery(), p_str: str = BodyOrQuery()):
    return {"p_int": p_int, "p_float": p_float, "p_str": p_str}
```

I'd be happy to work on implementing this as a PR, and was wondering if this feature was something you'd consider adding to the framework.  I've only looked in the codebase minimally, and I did start to work on this, but I don't yet have a good feeling for how feasible this even is, and wanted to gauge interest here before I worked on this any more.

### Describe alternatives you've considered

The best alternative I've come up with is the one shown in the example above.  

Alternatively, maybe there could be a way to define multiple path operators that respond to the same method (e.g. POST), but would have different signatures.  I don't really think this is viable, but figured I'd mention it.  I guess it would look like this:

```python
@app.post("/b_or_q")
def only_body(body: Item):
    return body


@app.post("/b_or_q")
def only_query(p_int: int, p_float: float, p_str: str):
    return {"p_int": p_int, "p_float": p_float, "p_str": p_str}
```

If this even worked, which it doesn't (it always matches the first one), I think it would still have the drawback that you couldn't mix and match body/query args for the same argument.

### Environment

* OS: Linux / macOS]:
* FastAPI Version [e.g. 0.3.0]: 0.60.1
* Python version: 3.8.3

**Comment by NoahCardoza** [comment]
Is this something that you'd like to support within the module and would accept a PR for @tiangolo?

**Comment by robswc** [comment]
Sorry to revive something this old.  Did this ever get added/considered?

**Comment by YuriiMotov** [comment]
Similar question with answer: https://github.com/fastapi/fastapi/discussions/7425
